### ISSUE TITLE
Help for web app

### ISSUE THREAD

**Issue by JimTR** [GameServerManagers/LinuxGSM#2970]
## User Story

As a web developer, I want to add gnu screen so that the web app works.

## Basic info

* **Distro:** [Ubuntu 18.04]
* **Game:** [All]
* **Command:** [status/start/stop/restart/console]
* **LinuxGSM version:** [v20.1.3]

## Further Information

I have a web interface for lgsm under development, however tmux does not 'play nice' with the web user even if this user is the same user as the lgsm user, but screen does. I have tried various methods to get tmux to work however when this achieved the are security issuses.
so the next method to try was to add screen functionality to lgsm which I have done. Currently my code will allow the user to choose either tmux or screen as the  multiplexer by adding a variable to the config files(s).
my knowledge of bash can be written on a postage stamp so is there anyone out there that wishes to review my code additions & tell me where the couple of bugs are so they can be fixed rather than doing requests into the project  

## Further Reading
none yet
Provide any useful documentation or resources that might help.

**Comment by dgibbs64** [comment]
@Grimston any thoughts on this as I know you are creating something as well.

**Comment by Grimston** [comment]
I have it working in my own project using tmux, as I have just implemented a full SSH web terminal.
I don't really see a need to move to screen, or even provide an option, as it should be trivial to just attach to the session or as I used to use, call: `tmux capture-pane` to get the current output on a timer.

Another option I looked at was to use the log file and stream that if changes are made.

**Comment by JimTR** [comment]
@Grimston  try running up apache using the suexec module to ensure you have the correct user then run php code like `<?php
if (php_sapi_name() == "cli") {
    // In cli-mode
define ('cr',PHP_EOL);
} else {
    // Not in cli-mode
define ('cr','<br>');
}
echo 'Running as user '.shell_exec('whoami').cr;
echo 'passthru '.cr; 
$output = passthru('tmux ls');
echo 'tmux output'.cr;
echo trim($output);
$output= passthru('screen -ls');
echo 'Screen Output'.cr;
echo trim($output);
echo PHP_EOL.'shell_exec'.cr;
$output = shell_exec('tmux ls');
echo 'Tmux Output'.cr;
echo trim($output);
$output = shell_exec('screen -ls');
echo 'screen Output'.cr;
echo trim($output);
 ?>
 ` 
from the the command line and via a browser you will see the difference .. if you can find the fix for this let me know

**Comment by thegreatstorm** [comment]
@JimTR If you want a webapp to interact with LinuxGSM I recommend using Python Flask. So you can just call the backend with rest api.
This is the app I been working on.
https://github.com/thegreatstorm/projectstorm

**Comment by JimTR** [comment]
Can not do that as such as the code I have written is a plugin for some other code so I guess the best option is to abandon LGSM and write the functions I need. I didn't want to do that but for about 20 lines of bash code to be added to the core that could be undocumented and not advertised this appears to be my only option.

**Comment by Grimston** [comment]
You could open an SSH connection using the correct user and use that to run the commands.

**Comment by JimTR** [comment]
Ok that would work however I am trying to 'keep things simple' rather than introducing more code that needs to be maintained.
The code as is sets the apache user to the game server user, so I am the game server user ... therefore I should be able to run commands as that user. everything works as expected with the exception of tmux. I guess this to be design feature of tmux rather than a miss configuration of the user. On the other hand gnu screen works as expected. After testing game servers over the last 6 months using screen as the multiplexer I can report that there are no adverse effects of using screen over using tmux. in fact it is easier as you can specify a log file when you open the window where tmux it **appears** you have to create the file [https://github.com/GameServerManagers/LinuxGSM/blob/master/lgsm/functions/command_start.sh](url) line 71.
looking on the net there are loads of posts saying that tmux is better than screen but to be fair this application is not using any of the 'advanced' functions of tmux I can not see the issue with 

1. adding a few lines of code to command_start.sh, command_stop.sh and check_status.sh to implement this
2. either making it known or not, to add a variable to one of the game servers config files to use screen over tmux  
3. the code I have written defaults to tmux so really what is the problem 
@Grimston 
> I don't really see a need to move to screen, or even provide an option

why say that ? does it not fit the app your building ?

**Comment by Grimston** [comment]
There is not much code for SSH to maintain as it's an extension for PHP already: [SSH2 Manual](https://www.php.net/manual/en/book.ssh2.php) or you can go for the full PHP route and use [phpseclib](http://phpseclib.sourceforge.net/)

As for tmux and the console log, you do not need to create the file, LinuxGSM outputs multiple things to the file, so it's created manually, tmux is told to pipe `>>` it's output to the file a little later on.

From what I understand you are trying to rely on screens ability to share sessions with other users, which in my mind is a security flaw and really should be avoided.

My reason for not needing screen is a full SSH connection to the correct user, so it really has little effect on me if it's tmux or screen. 
![SSH Console](https://cdn.discordapp.com/attachments/127498813903601664/733704282490142730/unknown.png)

I did notice you have some Python scripts in your repository, why not take advantage of Python for it? This is an [Old Project](https://github.com/Grimston/LinuxGamePanel) (Abandoned) that uses Python and provides a very basic WebUI for LinuxGSM.

**Comment by JimTR** [comment]
You have that totally wrong I am not sharing any screen windows with anyone who would do that as you point out that is a security risk. I have looked at my repository and can see no python scripts at all.
program flow as follows :
<b>Apache Request to start game server
php shell_exec or passthro starts the server in a multiplexer window
php then reads the console log file and processes the output which is then rendered to a web page
user then has the ability to send data to the multiplexer window via a form
currently that sent data is sanitised to only allow certain commands to sent to the multiplexer window
if successful the data is sent to multiplexer window and executed (tmux sendkeys) and the web page is updated</b></br>
   <img src="https://i.imgur.com/ZjXbZky.png"/>
The php code also requests auth to the page (login) and if you want to have a go try <a href="https://warmmail.co.uk/console.php?id=cssserver">here</a> , which I guess clears up how I'm using the script and if the security issues that may arise<br><br>
Why I wont use ssh is as follows I have a client that is on a server where they are not granted ssh access all they have is a file manager & cron manager ... I copied all the required game server files to their web space rather than installing the game via <server_name> i, got the supplier to add all dependences (that took a week or 2 before they would) then invoked the game via a php script all went well until the tmux issue switched to screen & problem solved. 
I have tried to get the client to host his games on one of my servers which are not as restrictive but he does not wish to do that even tho the result would be better, that said some one else in the same position
